src/devices/Arduino/ArduinoBoard.cs

```diff
@@ -0,0 +1,413 @@
+                    try
+                    {
+                        b = new ArduinoBoard(port, baud);
+                        b.Initialize();
+                        return b;
+                    }
+                    catch (Exception x) when (x is NotSupportedException || x is TimeoutException || x is IOException || x is UnauthorizedAccessException)
+                    {
+                        b?.Dispose();
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Searches all available com ports for an Arduino device.
+        /// </summary>
+        /// <returns>A board, already open and initialized. Null if none was found.</returns>
+        /// <remarks>
+        /// Scanning serial ports may affect unrelated devices. If there are problems, use the
+        /// <see cref="FindBoard(System.Collections.Generic.IEnumerable{string},System.Collections.Generic.IEnumerable{int})"/> overload excluding ports that shall not be tested.
+        /// </remarks>
+        public static ArduinoBoard? FindBoard()
+        {
+            return FindBoard(SerialPort.GetPortNames(), CommonBaudRates());
+        }
+
+        /// <summary>
+        /// Returns a list of commonly used baud rates.
+        /// </summary>
+        public static List<int> CommonBaudRates()
+        {
+            return new List<int>()
+            {
+                9600,
+                19200,
+                18400,
+                57600,
+                115200,
+                230400,
+                250000,
+                500000,
+                1000000,
+                2000000,
+            };
+        }
+
+        /// <summary>
+        /// Initialize the board connection. This must be called before any other methods of this class.
+        /// </summary>
+        /// <exception cref="NotSupportedException">The Firmata firmware on the connected board is too old.</exception>
+        public virtual void Initialize()
+        {
+            if (_firmata != null)
+            {
+                throw new InvalidOperationException("Board already initialized");
+            }
+
+            _firmata = new FirmataDevice();
+            _firmata.Open(_dataStream);
+            _firmata.OnError += FirmataOnError;
+            FirmataVersion = _firmata.QueryFirmataVersion();
+            if (FirmataVersion < _firmata.QuerySupportedFirmataVersion())
+            {
+                throw new NotSupportedException($"Firmata version on board is {FirmataVersion}. Expected {_firmata.QuerySupportedFirmataVersion()}. They must be equal.");
+            }
+
+            Log($"Firmata version on board is {FirmataVersion}.");
+
+            FirmwareVersion = _firmata.QueryFirmwareVersion(out var firmwareName);
+            FirmwareName = firmwareName;
+
+            Log($"Firmware version on board is {FirmwareVersion}");
+
+            _firmata.QueryCapabilities();
```

Comment:
Looks like this isn't actually doing anything with the state other than testing if communication is working right? If this is just for debugging purposes, it might be better to add this code and the one that actually logs all of the capabilities  behind #Debug flag to not affect perf. You don't have to do this, I'm just mentioning it for you to consider.

Reply:
No, this is filling the capability list (see next line). So this is required at startup. The log part will later be routed trough the log library, with a LogLevel of debug or info.